src/backends/kafka/mod.rs

```diff
@@ -178,30 +189,56 @@ impl<'a> ArroyoConsumer<'a, KafkaPayload> for KafkaConsumer {
         &mut self,
         timeout: Option<Duration>,
     ) -> Result<Option<ArroyoMessage<KafkaPayload>>, ConsumerError> {
-        let duration = timeout.unwrap_or(Duration::from_millis(100));
+        self.state.assert_consuming_state()?;
 
-        match self.consumer.as_mut() {
-            None => Err(ConsumerError::ConsumerClosed),
-            Some(consumer) => {
-                let res = consumer.poll(duration);
-                match res {
-                    None => Ok(None),
-                    Some(res) => {
-                        let msg = res?;
-                        Ok(Some(create_kafka_message(msg)))
-                    }
-                }
+        let duration = timeout.unwrap_or(Duration::from_millis(100));
+        let consumer = self.consumer.as_mut().unwrap();
+        let res = consumer.poll(duration);
+        match res {
+            None => Ok(None),
+            Some(res) => {
+                let msg = res?;
+                Ok(Some(create_kafka_message(msg)))
             }
         }
     }
 
-    fn pause(&mut self, _: HashSet<Partition>) -> Result<(), ConsumerError> {
-        //TODO: Implement this
+    fn pause(&mut self, partitions: HashSet<Partition>) -> Result<(), ConsumerError> {
+        self.state.assert_consuming_state()?;
+
+        let mut topic_map = HashMap::new();
+        for partition in partitions {
+            let offset = *self
```

Comment:
Why did you need to explicitly dereference self here ?

Reply:
HashMap::get() on `self.offsets` returns an Option<&u64>. We copy the offset number into our new `topic_map`.